modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/AbstractExternalResult.java

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.processors.query.h2.H2MemoryTracker;
+import org.apache.ignite.internal.processors.query.h2.H2Utils;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.result.ResultExternal;
+import org.h2.result.ResultInterface;
+import org.h2.store.Data;
+import org.h2.value.Value;
+import org.jetbrains.annotations.NotNull;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing.DISK_SPILL_DIR;
+
+/**
+ * Basic class for external result. Contains common methods for file IO.
+ */
+@SuppressWarnings({"MissortedModifiers", "WeakerAccess", "ForLoopReplaceableByForEach"})
+public abstract class AbstractExternalResult implements ResultExternal {
+    /** File name generator. */
+    private static final AtomicLong idGen = new AtomicLong();
+
+    /** Serialized row header size. */
+    protected static final int ROW_HEADER_SIZE = 8; // 2 x int: row length in bytes + column count.
+
+    /** Logger. */
+    protected final IgniteLogger log;
+
+    /** Current size in rows. */
+    protected int size;
+
+    /** File. */
+    protected final File file;
+
+    /** Spill file IO.*/
+    protected volatile FileIO fileIo;
+
+    /** Memory tracker. */
+    protected final H2MemoryTracker memTracker;
+
+    /** Parent result. */
+    protected final AbstractExternalResult parent;
```

Comment:
This construction is needed for query results caching. Sometimes H2 caches the query results instead of running a new query. And for this cases child results are created. These child instances share the same file for reading the result. Despite multiple children can be created and each this child is used by it's own thread, this construction looks thread safe because file is effectively final - no more data can be written to it when child is allowed to be created. When we've done with all children, the underlying file should be closed - so we need to have some bookkeeping of the results which use the same offload file.

Reply:
But it looks like the same the same FileIO shouldn't be shared between all children - each child should open it's own FileIO. I'll fix it.